### PR TITLE
Fluent configuration

### PR DESCRIPTION
**Pull request by michael-rubel**
By using objects in the `ServiceProvider`, we will be able to define automatic bindings for folders using fluent methods rather than using a messy configuration file.

We can now do this:
```php
AutoBinder::from(folder: 'Services')
    ->as('singleton')
    ->bind();
```

We are no longer constrained to subfolders of implementation classes, and you will be able to place your interfaces wherever you like.

```php
AutoBinder::from(folder: 'Services')
    ->basePath('app')
    ->classNamespace('App')
    ->interfaceNamespace('Interfaces')
    ->as('singleton')
    ->bind();
```

If the interface namespace is not defined explicitly, the package will guess the interface folder based on the class folder and class namespace. For example:

```php
AutoBinder::from(folder: 'Services')
    ->basePath('app/Domain')
    ->classNamespace('App\\Domain\\Services')
    ->interfaceNamespace('App\\Domain\\Services\\Interfaces')
    ->as('singleton')
    ->bind();
```

### CODE REVIEW

**Comment by michael-rubel** [comment]
Added the possibility to map over multiple folders at once:

```php
AutoBinder::from('Services', 'Models')->each(
    fn ($binder) => $binder->basePath('tests/Boilerplate')
        ->classNamespace('MichaelRubel\\AutoBinder\\Tests\\Boilerplate')
        ->interfaceNamespace("MichaelRubel\\AutoBinder\\Tests\\Boilerplate\\$binder->classFolder\\Interfaces")
        ->as('singleton')
        ->bind()
);
```

**Comment by michael-rubel** [comment]
Added possibility to set interface name used by the project:
```php
AutoBinder::from(folder: 'Services')
    ->interfaceNaming('Contract')
    ->bind();
```

**Comment by michael-rubel** [comment]
Added the `when` method to manually configure class dependencies if the binder encountered them during scanning:

```php
AutoBinder::from(folder: 'Services')
    ->when(ExampleService::class, function ($app, $service) {
        return new ExampleService(true);
    })
    ->when(AnotherServiceInterface::class, function ($app, $service) {
        return new AnotherService(true);
    })
    ->bind();
```

You can pass interface or concrete class to `when` as the first parameter. The interface has a higher priority.

**Comment by michael-rubel** [comment]
Added the possibility to ignore subdirectories:
```php
AutoBinder::from(folder: 'Services')
    ->exclude('Traits', 'Contracts')
    ->bind();
```

File: src/AutoBinder.php [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+     * Create the object with target folder assigned.
+     *
+     * @param string|array $folder
+     *
+     * @return static|Collection
+     */
+    public static function from(string|array $folder): static|Collection
+    {
+        if (func_num_args() > 1) {
+            $folders = is_array($folder) ? $folder : func_get_args();
```
Comment (olsza) [inline src/AutoBinder.php:93]: refactoring ;)
```suggestion
            $folders = is_array($folder)
            ? $folder
            : func_get_args();
```

File: src/AutoBinder.php [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+            return collect($folders)->map(fn ($folder) => new static($folder));
+        }
+
+        $folder = is_string($folder) ? $folder : current($folder);
```
Comment (olsza) [inline src/AutoBinder.php:98]: refactoring
```suggestion
        $folder = is_string($folder)
        ? $folder
        : current($folder);
```

File: src/AutoBinder.php [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    /**
+     * Exclude specified subdirectory from scanning.
+     *
+     * @param string|array $folders
+     *
+     * @return static
+     */
+    public function exclude(string|array $folders): static
+    {
+        $folders = is_array($folders) ? $folders : func_get_args();
```
Comment (olsza) [inline src/AutoBinder.php:112]: refactoring
```suggestion
        $folders = is_array($folders)
        ? $folders
        : func_get_args();
```

**Review by olsza — APPROVED** [review]

File: src/AutoBinder.php (hunk 1 above)
Comment (michael-rubel) [reply src/AutoBinder.php:93]: Thanks, but I prefer to stick with a one-liner. 🙂

File: src/Traits/BindsToContainer.php [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+     * Build the interface class-string based on the class folder.
+     *
+     * @param string $filename
+     *
+     * @return string
+     */
+    protected function buildInterfaceFromClassBy(string $filename): string
+    {
+        return $this->classNamespace . '\\'
+            . $this->classFolder . '\\'
```
Comment (valeriisotnyk) [inline src/Traits/BindsToContainer.php:154]: Maybe this 2 lines  can move to some method. Duplicate

File: src/Traits/BindsToContainer.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+     *
+     * @param string $filename
+     *
+     * @return string
+     */
+    protected function buildInterfaceBy(string $filename): string
+    {
+        return $this->interfaceNamespace . '\\'
+            . $this->prepareNamingFor($filename)
+            . ($this->interfaceNaming);
```
Comment (valeriisotnyk) [inline src/Traits/BindsToContainer.php:141]: Why there you have `(` `)` ?

File: src/Traits/BindsToContainer.php (hunk 4 above)
Comment (michael-rubel) [reply src/Traits/BindsToContainer.php:154]: I think it's more readable to leave it as it is.
Additional method is overhead in this case.

File: src/Traits/BindsToContainer.php (hunk 5 above)
Comment (michael-rubel) [reply src/Traits/BindsToContainer.php:141]: You're right, they're actually redundant. Thanks!

**Pull request merged**